src/circular_buffer.rs

```diff
@@ -0,0 +1,366 @@
+
+        let mut new_buffer = Vec::with_capacity(new_capacity);
+        new_buffer.resize(new_capacity, 0);
+
+        if self.length > 0 {
+            if self.position + self.length <= self.current_capacity() {
+                new_buffer[..self.length]
+                    .copy_from_slice(&self.buffer[self.position..self.position + self.length]);
+            } else {
+                let a = self.position..self.current_capacity();
+                let b = 0..self.length - a.len();
+                new_buffer[..a.len()].copy_from_slice(&self.buffer[a.clone()]);
+                new_buffer[a.len()..a.len() + b.len()].copy_from_slice(&self.buffer[b]);
+            }
+        }
+
+        self.buffer = new_buffer.into_boxed_slice();
+        self.position = 0;
+    }
+
+    #[inline(never)]
+    #[cold]
+    fn grow_buffer(&mut self) -> bool {
+        if self.remaining_mut() == 0 {
+            return false;
+        }
+
+        let new_capacity = std::cmp::min(
+            self.max_capacity,
+            std::cmp::max(self.current_capacity() * 2, MINIMUM_NON_EMPTY_CAPACITY),
+        );
+
+        self.resize_buffer(new_capacity);
+        true
+    }
+
+    fn advance_mut_impl(&mut self, count: usize) {
+        assert!(
+            count <= self.remaining_mut(),
+            "tried to advance the write cursor past maximum buffer capacity"
+        );
+        assert!(
+            count <= self.remaining_mut_without_realloc(),
+            "tried to advance the write cursor past what was written"
+        );
+        self.length += count;
+    }
+
+    fn bytes_mut_impl(&mut self) -> &mut [u8] {
+        if self.remaining_mut_without_realloc() == 0 {
+            if !self.grow_buffer() {
+                return &mut self.buffer[..0];
+            }
+        }
+
+        let position_mut = (self.position + self.length) % self.current_capacity();
+        let range = position_mut
+            ..std::cmp::min(
+                position_mut + self.remaining_mut_without_realloc(),
+                self.current_capacity(),
+            );
+        &mut self.buffer[range]
+    }
+
+    pub fn read_cursor(&self) -> (usize, usize) {
+        (self.position, self.length)
+    }
+
+    pub fn set_read_cursor(&mut self, (position, length): (usize, usize)) {
+        self.position = position;
+        self.length = length;
+    }
+
+    pub fn read_exact_into_vec(&mut self, length: usize) -> Vec<u8> {
+        assert!(length <= self.remaining());
+        let mut output = Vec::with_capacity(length);
+        let first_chunk_size = std::cmp::min(length, self.bytes().len());
+        output.extend_from_slice(&self.bytes()[..first_chunk_size]);
+        self.advance(first_chunk_size);
+
+        let remaining = length - first_chunk_size;
+        output.extend_from_slice(&self.bytes()[..remaining]);
+        self.advance(remaining);
+
+        output
+    }
+
+    pub fn apply_soft_limit(&mut self, limit: usize) {
```

Comment:
I haven't looked deeply into the usage of it, but is it possible to abuse (doing O(n) work with O(1) ops) this way (going back and forth between soft_limit and 2x soft_limit):
```rust
    #[test]
    fn abuse_soft_limit() {
        let mut b = CircularBuffer::new(0, 16);
        b.write_all(b"012345678").unwrap();
        b.apply_soft_limit(8);
        assert_eq!(b.current_capacity(), 16);
        b.advance(1);
        b.apply_soft_limit(8);
        assert_eq!(b.current_capacity(), 8);
        b.write_all(b"0").unwrap();
        assert_eq!(b.current_capacity(), 16);
        b.advance(1);
        b.apply_soft_limit(8);
        assert_eq!(b.current_capacity(), 8);
    }
```

Reply:
Hmm... you're right, someone could conceivably write a bunch of really tiny messages to the buffer so that the last message barely crosses the threshold, wait until it's processed, send another one, and repeat.
 
Although technically it's *not* O(n), it's still O(1) as the maximum amount of bytes copied will be equal to the soft cap limit (so it's effectively constant, even though the size of the constant factors significantly differs). (:

I'd rather not remove it though, as this *is* quite important in keeping the memory usage down in case the buffers are unintentionally bloated.

So off the top of my head here's what I think we could do:

1) Split the buffer into two - soft capped part and hard capped part, and just never reallocate the soft capped part.
2) Only apply the soft-cap once there is no data in the buffer. If there's no data then there's no need to copy anything.

I'd probably go for (2) since it's simpler. If this library is going to be eventually replaced anyway as @grbIzl said then I'd rather keep the complexity down to a minimum.

So what do you think?

Reply:
I agree, option (2) sounds good to me.

As for the replacement, @grbIzl are you talking about https://github.com/paritytech/soketto? I don't see how it does memory limiting, only that https://github.com/paritytech/soketto/blob/042e5b74de6b3324994c1754d525b2fed5195db8/src/connection.rs#L27 cc @tomaka @niklasad1.

Anyway, I don't think it's going to be replaced anytime soon.

Reply:
> I agree, option (2) sounds good to me.

Great! So that's what I already did in the most recent push, if you care to take a look. (:

(It was a `--force` push since I had to get rid of the version bump and sign the commits, but the old commits are unchanged.)

Reply:
> Anyway, I don't think it's going to be replaced anytime soon.

I hope to get the migration done as quickly as possible and definitely this year.

Reply:
Here's our current thinking about resource limiting: https://github.com/paritytech/jsonrpsee/issues/203 – this thread here and @koute 's findings are very relevant for us to ensure we build the right thing (so no wasted work!).